bson/primitive/primitive_test.go

```diff
@@ -199,3 +200,334 @@ func TestDateTime(t *testing.T) {
+			err := json.Unmarshal(b, &got)
+			assert.NoError(t, err)
+			assert.Equal(t, tc.test, got)
+		})
+	}
+
+	for _, tc := range []struct {
+		name string
+		test string
+	}{
+		{
+			"illegal",
+			`nil`,
+		},
+		{
+			"invalid",
+			`[null, 3.142ipsum]`,
+		},
+		{
+			"malformatted",
+			`[null, 3.142,]`,
+		},
+		{
+			"truncated",
+			`[2.71828, "Lorem ipsum", null,`,
+		},
+		{
+			"object type",
+			`{"pi": 3.142}`,
+		},
+		{
+			"number type",
+			`3.142`,
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			var a []interface{}
+			want := json.Unmarshal([]byte(tc.test), &a)
+			var b A
+			got := json.Unmarshal([]byte(tc.test), &b)
+			switch w := want.(type) {
+			case *json.UnmarshalTypeError:
+				w.Type = reflect.TypeOf(b)
+				require.IsType(t, want, got)
+				g := got.(*json.UnmarshalTypeError)
+				assert.Equal(t, w, g)
+			default:
+				assert.Equal(t, want, got)
+			}
+		})
+	}
+}
+
+func TestUnmarshalingM(t *testing.T) {
+	t.Parallel()
+
+	for _, tc := range []struct {
+		name string
+		test M
+	}{
+		{
+			"nil",
+			nil,
+		},
+		{
+			"empty",
+			M{},
+		},
+		{
+			"non-empty",
+			M{
+				"a": "Lorem ipsum",
+				"b": 3.142,
+				"c": true,
+				"d": nil,
+				"f": A{"Lorem ipsum", 3.142, true, nil, M{"Lorem": "ipsum"}},
+				"g": M{"foo": "bar"},
+			},
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			b, _ := json.Marshal(tc.test)
+			var got M
+			err := json.Unmarshal(b, &got)
+			assert.NoError(t, err)
+			assert.Equal(t, tc.test, got)
+		})
+	}
+
+	for _, tc := range []struct {
+		name string
+		test string
+	}{
+		{
+			"illegal",
+			`nil`,
+		},
+		{
+			"invalid",
+			`{"pi": 3.142ipsum}`,
+		},
+		{
+			"malformatted",
+			`{"pi", 3.142}`,
+		},
+		{
+			"truncated",
+			`{"pi": 3.142`,
+		},
+		{
+			"array type",
+			`["pi", 3.142]`,
+		},
+		{
+			"string type",
+			`"pi"`,
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			var a map[string]interface{}
+			want := json.Unmarshal([]byte(tc.test), &a)
+			var b M
+			got := json.Unmarshal([]byte(tc.test), &b)
+			switch w := want.(type) {
+			case *json.UnmarshalTypeError:
+				w.Type = reflect.TypeOf(b)
+				require.IsType(t, want, got)
+				g := got.(*json.UnmarshalTypeError)
+				assert.Equal(t, w, g)
+			default:
+				assert.Equal(t, want, got)
+			}
+		})
+	}
+}
+
+func TestUnmarshalingD(t *testing.T) {
+	t.Parallel()
+
+	for _, tc := range []struct {
+		name string
+		test D
+	}{
+		{
+			"nil",
+			nil,
+		},
+		{
+			"empty",
+			D{},
+		},
+		{
+			"non-empty",
+			D{
+				{"hello", "world"},
+				{"pi", 3.142},
+				{"boolean", true},
+				{"nothing", nil},
+				{"list", A{"hello world", 3.142, false, nil, D{{"Lorem", "ipsum"}}}},
+				{"document", D{{"foo", "bar"}}},
+			},
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			b, _ := json.Marshal(tc.test)
+			var got D
+			err := json.Unmarshal(b, &got)
```

Comment:
Testing that JSON unmarshaling works with a document marshaled by the custom JSON marshaler can hide bugs in the custom JSON unmarshaling logic. For example, if the custom JSON marshaling code always uses a semicolon instead of a colon for a key/value separator, the JSON unmarshaling code could incorrectly expect a semicolon separator, but the tests would still pass.

We should use a known-correct JSON document (e.g. a string literal, embedded string from a text doc, or generated with a trusted reference) to test the JSON unmarshaling logic. That recommendation applies to all `Unmarshal` tests, but is most relevant here because both marshaling and unmarshaling are custom.